File: swig/R/test_decomp.R

test_SparseProject <- function () {
  set.seed(0)
  X = matrix(rnorm(20000 * 100),nrow = 20000,ncol = 100,byrow = FALSE)
  X = X / matrix(rep(sqrt(colSums(X*X)),nrow(X)),nrow(X),ncol(X),byrow=T)
  param = list('numThreads' = -1, # number of processors/cores to use (-1 => all cores)
             'pos' = FALSE,
             'mode'= 1, # projection on the l1 ball
             'thrs' = 2)
  .printf( "\n  Projection on the l1 ball\n")
  tic = proc.time()
  X1 = spams.SparseProject(X,param)
  tac = proc.time()
  t = (tac - tic)[['elapsed']]
  .printf("  Time : %f\n", t)
  if (t != 0)
    .printf("%f signals of size %d projected per second\n",(ncol(X) / t),nrow(X))
  s = colSums(abs(X1))
  .printf("Checking constraint: %f, %f\n",min(s),max(s))

  .printf("\n  Projection on the Elastic-Net\n")
  param['mode'] = 2  # projection on the Elastic-Net
  param['lambda1'] = 0.15
  tic = proc.time()
  X1 = spams.SparseProject(X,param)
  tac = proc.time()
  t = (tac - tic)[['elapsed']]
  .printf("  Time : %f\n", t)
  if (t != 0)
    .printf("%f signals of size %d projected per second\n",(ncol(X) / t),nrow(X))
  print(param['lambda1'])
  constraints = colSums(X1*X1) + param[['lambda1']] * colSums(abs(X1))
  .printf("Checking constraint: %f, %f (Projection is approximate : stops at a kink)\n",min(constraints),max(constraints))

  .printf("\n  Projection on the FLSA\n")
  param['mode'] = 6       # projection on the FLSA
  param['lambda1'] = 0.7
  param['lambda2'] = 0.7
  param['lambda3'] = 1.0
  X = matrix(runif(2000 * 100,0,1),nrow = 2000,ncol = 100,byrow = FALSE)
  X = X / matrix(rep(sqrt(colSums(X*X)),nrow(X)),nrow(X),ncol(X),byrow=T)
 # matlab : X=X./repmat(sqrt(sum(X.^2)),[size(X,1) 1]);
  tic = proc.time()
  X1 = spams.SparseProject(X,param)
  tac = proc.time()
  t = (tac - tic)[['elapsed']]
  .printf("  Time : %f\n", t)
  if (t != 0)
    .printf("%f  signals of size %d projected per second\n",ncol(X) / t,nrow(X))
#  constraints = 0.5 * param[['lambda3']] * colSums(X1*X1) + param[['lambda1']] * colSums(abs(X1)) + param[['lambda2']] * abs(X1[3:nrow(X1),]) - colSums(X1[2,])
  constraints = 0.5 * param[['lambda3']] * colSums(X1*X1) + param[['lambda1']] * colSums(abs(X1)) + param[['lambda2']] * abs(X1[3:nrow(X1),])  - colSums(matrix(X1[2,],nrow=1))
  .printf("Checking constraint: %f, %f (Projection is approximate : stops at a kink)\n",min(constraints),max(constraints))
  return(NULL)
}

test_Lasso <- function() {
  set.seed(0)
  .printf("test Lasso\n")
##############################################
# Decomposition of a large number of signals
##############################################
# data generation
  X = matrix(rnorm(100 * 100000),nrow = 100,ncol = 100000,byrow = FALSE)
  X = X / matrix(rep(sqrt(colSums(X*X)),nrow(X)),nrow(X),ncol(X),byrow=T)
  D = matrix(rnorm(100 * 200),nrow = 100,ncol = 200,byrow = FALSE)
  D = D / matrix(rep(sqrt(colSums(D*D)),nrow(D)),nrow(D),ncol(D),byrow=T)
  param = list( 'lambda' = 0.15, # not more than 20 non-zeros coefficients
    'numThreads' = -1, # number of processors/cores to use; the default choice is -1
                       # and uses all the cores of the machine
        'mode' = 'PENALTY')        # penalized formulation
  tic = proc.time()
  alpha = spams.Lasso(X,D,param,return_reg_path = FALSE)
  tac = proc.time()
  t = (tac - tic)[['elapsed']]
  .printf("%f signals processed per second\n",as.double(ncol(X)) / t)
########################################
# Regularization path of a single signal 
########################################
  X = matrix(rnorm(64),nrow = 64,ncol = 1,byrow = FALSE)
  D = matrix(rnorm(64 * 10),nrow = 64,ncol = 10,byrow = FALSE)
  D = D / matrix(rep(sqrt(colSums(D*D)),nrow(D)),nrow(D),ncol(D),byrow=T)
  res = spams.Lasso(X,D,param,return_reg_path = TRUE)
  alpha = res[[1]]
  path = res[[2]]
  return(NULL)
}


#
test_decomp.tests =list( 
  'SparseProject' = test_SparseProject,
  'Lasso' = test_Lasso)

File: swig/python/test_decomp.py
import sys
import numpy as np
import scipy
import scipy.sparse

import spams
import time
from test_utils import *

if not ('rand' in scipy.sparse.__dict__):
    import myscipy_rand as ssp
else:
    import scipy.sparse as ssp

def test_sparseProject():
    np.random.seed(0)
    X = np.asfortranarray(np.random.normal(size = (20000,100)))
    # matlab : X=X./repmat(sqrt(sum(X.^2)),[size(X,1) 1]);
    X = np.asfortranarray(X / np.tile(np.sqrt((X*X).sum(axis=0)),(X.shape[0],1)))
    param = {'numThreads' : -1, # number of processors/cores to use (-1 => all cores)
             'pos' : False,
             'mode': 1, # projection on the l1 ball
             'thrs' : 2}
    print "\n  Projection on the l1 ball"
    tic = time.time()
    X1 = spams.sparseProject(X,**param)
    tac = time.time()
    t = tac - tic
    print "  Time : ", t
    if (t != 0):
        print "%f signals of size %d projected per second" %((X.shape[1] / t),X.shape[0])
    s = np.abs(X1).sum(axis=0)
    print "Checking constraint: %f, %f" %(min(s),max(s))

    print "\n  Projection on the Elastic-Net"
    param['mode'] = 2  # projection on the Elastic-Net
    param['lambda1'] = 0.15
    tic = time.time()
    X1 = spams.sparseProject(X,**param)
    tac = time.time()
    t = tac - tic
    print "  Time : ", t
    if (t != 0):
        print "%f signals of size %d projected per second" %((X.shape[1] / t),X.shape[0])
    constraints = (X1*X1).sum(axis=0) + param['lambda1'] * np.abs(X1).sum(axis=0)
    print 'Checking constraint: %f, %f (Projection is approximate : stops at a kink)' %(min(constraints),max(constraints))
    
    print "\n  Projection on the FLSA"
    param['mode'] = 6       # projection on the FLSA
    param['lambda1'] = 0.7
    param['lambda2'] = 0.7
    param['lambda3'] = 1.0
    X = np.asfortranarray(np.random.random(size = (2000,100)))
    # matlab : X=X./repmat(sqrt(sum(X.^2)),[size(X,1) 1]);
    X = np.asfortranarray(X / np.tile(np.sqrt((X*X).sum(axis=0)),(X.shape[0],1)))
    tic = time.time()
    X1 = spams.sparseProject(X,**param)
    tac = time.time()
    t = tac - tic
    print "  Time : ", t
    if (t != 0):
        print "%f signals of size %d projected per second" %((X.shape[1] / t),X.shape[0])
    constraints = 0.5 * param['lambda3'] * (X1*X1).sum(axis=0) + param['lambda1'] * np.abs(X1).sum(axis=0) + \
    param['lambda2'] * np.abs(X1[2:,] - X1[1:-1,]).sum(axis=0)
    print 'Checking constraint: %f, %f (Projection is approximate : stops at a kink)' %(min(constraints),max(constraints))
    return None

def test_cd():
    return None

def test_L1L2BCD():
    return None

def test_lasso():
    np.random.seed(0)
    print "test Lasso"
##############################################
# Decomposition of a large number of signals
##############################################
# data generation
    X = np.asfortranarray(np.random.normal(size=(100,100000)))
    # X=X./repmat(sqrt(sum(X.^2)),[size(X,1) 1]);
    X = np.asfortranarray(X / np.tile(np.sqrt((X*X).sum(axis=0)),(X.shape[0],1)))
    D = np.asfortranarray(np.random.normal(size=(100,200)))
    D = np.asfortranarray(D / np.tile(np.sqrt((D*D).sum(axis=0)),(D.shape[0],1)))
    # parameter of the optimization procedure are chosen
#param.L=20; # not more than 20 non-zeros coefficients (default: min(size(D,1),size(D,2)))
    param = {
        'lambda1' : 0.15, # not more than 20 non-zeros coefficients
        'numThreads' : -1, # number of processors/cores to use; the default choice is -1
        # and uses all the cores of the machine
        'mode' : 2}        # penalized formulation

    tic = time.time()
    alpha = spams.lasso(X,D = D,return_reg_path = False,**param)
    tac = time.time()
    t = tac - tic
    print "%f signals processed per second\n" %(float(X.shape[1]) / t)
########################################
# Regularization path of a single signal 
########################################
    X = np.asfortranarray(np.random.normal(size=(64,1)))
    D = np.asfortranarray(np.random.normal(size=(64,10)))
    D = np.asfortranarray(D / np.tile(np.sqrt((D*D).sum(axis=0)),(D.shape[0],1)))
    (alpha,path) = spams.lasso(X,D = D,return_reg_path = True,**param)
    return None

def test_lassoMask():
    return None

def test_lassoWeighted():
    return None

def test_omp():
    return None

def test_ompMask():
    return None

def test_somp():
    return None


tests = {
    'sparseProject' : test_sparseProject,
    'cd' : test_cd,
    'L1L2BCD' : test_L1L2BCD,
    'lasso' : test_lasso,
    'lassoMask' : test_lassoMask,
    'lassoWeighted' : test_lassoWeighted,
    'omp' : test_omp,
    'ompMask' : test_ompMask,
    'somp' : test_somp,
    }
